feat(App): add global `portal` prop

menthol opened the pull request:
### 🔗 Linked issue

Resolves #3687

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Introduced a `portalTarget` option in `UApp` to specify portal mounting targets across various components. Updated portals in components like `ContextMenu`, `Drawer`, `Tooltip`, and others to respect `portalTarget`. Default target is set to `body`.

### 📝 Checklist
- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

benjamincanac reviewed (changes requested):
I'm not sure I would put this config in global `app.config` but instead put it in the `portal` prop of each component as an object with the `to` and `disabled` fields. It would be more consistent and would allow configuration per component.

menthol commented:
I've tried, and Implementing a system using a component-based configuration proved to be significantly more challenging than anticipated.  The process required meticulous definition of the target for each individual component, leading to a highly time-consuming and cumbersome configuration process.  This level of detail, while potentially offering great flexibility, introduces a substantial overhead.

A potential solution to mitigate this complexity might involve a hybrid approach: employing a global configuration file to establish baseline settings, while allowing for component-specific overrides where necessary. This would reduce the repetitive nature of defining targets for every component, simplifying the overall configuration management.

benjamincanac commented:
Then I would still not put it in App Config but as a prop on `UApp` that is provided to all components maybe?

menthol commented:
This is a great idea.

menthol commented:
Is this the direction you had in mind?

sebimalwieder commented:
Thanks @menthol for pushing this! Really helpful

benjamincanac reviewed (commented):
Maybe we could create a `usePortal` composable to avoid code duplication that also exports its symbol like: https://github.com/nuxt/ui/blob/v3/src/runtime/composables/useAvatarGroup.ts#L4

WDYT?

menthol commented:
> Thanks @menthol for pushing this! Really helpful

You're welcome. I'm waiting for it to be reviewed to see if it's suitable. I hope it will be merged for 3.0.3 or 3.0.4.

menthol committed:
feat: enhance portal support with flexible target options

Introduced `usePortal` composable to standardize portal handling across components, supporting `boolean`, `string`, and `HTMLElement` types. Updated various components to leverage this composable, improving configurability and integration with custom portal targets. Default behavior remains backward compatible.

menthol commented:
@benjamincanac What do you think of the changes? (i'm fixing the ci, but overall I think it's ready for review)

benjamincanac reviewed (commented):
It looks good! Maybe we could improve it even further by returning `target` and `to` so you could do: `const portalProps = usePortal(toRef(() => props.portal))` grouped with the `rootProps`, `contentProps`, etc. and `v-bind="portalProps"`.

benjamincanac commented on src/runtime/components/App.vue:
```diff
@@ -6,6 +6,7 @@ export interface AppProps<T extends Messages = Messages> extends Omit<ConfigProv
   tooltip?: TooltipProviderProps
   toaster?: ToasterProps | null
   locale?: Locale<T>
+  portalTarget?: string | HTMLElement
```
Should we just call it `portal` here to be consistent with other components?

menthol commented:
I've an error with v-bind, TypeScript want readonly bindings. Have no clues for now.

menthol commented:
@benjamincanac done

benjamincanac committed:
Merge branch 'v3' into v3

benjamincanac committed:
up

benjamincanac commented:
@menthol I've cleanup up the code a bit, can you confirm that it's good to you? 😊

menthol commented:
@benjamincanac 
LGTM,
but removing the computed in usePortal, you make the return not reactive anymore

menthol commented on src/runtime/composables/usePortal.ts:
```diff
@@ -0,0 +1,24 @@
+
+  provide(portalTargetInjectionKey, computed(() => to.value))
+
+  return {
+    to: to.value,
+    disabled: disabled.value
+  }
```
Maybe removing the .value access from the composable may prevent the reactivity from vanishing ? 

```suggestion
  return {
    to,
    disabled
  }
```

benjamincanac replied:
You're right, I've put back the `computed` on top!

benjamincanac committed:
up

benjamincanac commented:
Thanks @menthol! 😊

benjamincanac commented on src/runtime/composables/usePortal.ts:
```diff
@@ -0,0 +1,24 @@
+  const portalTarget = inject(portalTargetInjectionKey, undefined)
+
+  const to = computed(() => {
+    if (typeof portal.value === 'string' || portal.value instanceof HTMLElement) {
```
@menthol It seems `instanceof HTMLElement` breaks the build, can we check on `typeof portal.value === 'boolean'` instead? 🤔

benjamincanac replied:
![CleanShot 2025-04-22 at 10 54 31@2x](https://github.com/user-attachments/assets/d4a15b3e-6b7c-4ed0-a8ed-deba4cdc71a1)

menthol replied:
We can remove the HTMLElement check, was only for feature parity with Radix.

string and boolean work perfectly alone

menthol replied:
If the only problem is this `instanceof`, we can invert the condition

menthol commented on src/runtime/composables/usePortal.ts:
```diff
@@ -0,0 +1,24 @@
+export function usePortal(portal: Ref<string | HTMLElement | boolean | undefined>) {
+  const portalTarget = inject(portalTargetInjectionKey, undefined)
+
+  const to = computed(() => {
+    if (typeof portal.value === 'string' || portal.value instanceof HTMLElement) {
+      return portal.value
+    }
+
+    return portalTarget?.value ?? 'body'
+  })
```
```typescript
import { inject, provide, computed, type Ref, type InjectionKey } from 'vue'

export const portalTargetInjectionKey: InjectionKey<Ref<string | HTMLElement>> = Symbol('nuxt-ui.portal-target')

export function usePortal(portal: Ref<string | HTMLElement | boolean | undefined>) {
  const portalTarget = inject(portalTargetInjectionKey, undefined)

  const to = computed(() => {
    if (typeof portal.value === 'boolean' || portal.value === undefined) {
      return portalTarget?.value ?? 'body'
    }

    return portal.value
  })

  const disabled = computed(() => typeof portal.value === 'boolean' ? !portal.value : false)

  provide(portalTargetInjectionKey, computed(() => to.value))

  return computed(() => ({
    to: to.value,
    disabled: disabled.value
  }))
}
```

menthol replied:
New PR: #3954
